src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.GeneratedFileReplacingCompilationTracker.cs

```diff
@@ -0,0 +1,217 @@
+                if (this.ProjectState.LanguageServices == fromProject.LanguageServices)
+                {
+                    return compilation.ToMetadataReference(projectReference.Aliases, projectReference.EmbedInteropTypes);
+                }
+                else
+                {
+                    var version = await GetDependentSemanticVersionAsync(solution, cancellationToken).ConfigureAwait(false);
+                    return MetadataOnlyReference.GetOrBuildReference(solution, projectReference, compilation, version, cancellationToken);
+                }
+            }
+
+            public CompilationReference? GetPartialMetadataReference(ProjectState fromProject, ProjectReference projectReference)
+            {
+                // This method is used if you're forking a solution with partial semantics, and used to quickly produce references.
+                // So this method should only be called if:
+                //
+                // 1. Project A has a open source generated document, and this CompilationTracker represents A
+                // 2. Project B references that A, and is being frozen for partial semantics.
+                //
+                // We generally don't use partial semantics in a different project than the open file, so this isn't a scenario we need to support.
+                throw new NotImplementedException();
+            }
+
+            public async ValueTask<TextDocumentStates<SourceGeneratedDocumentState>> GetSourceGeneratedDocumentStatesAsync(SolutionState solution, CancellationToken cancellationToken)
+            {
+                var underlyingGeneratedDocumentStates = await _underlyingTracker.GetSourceGeneratedDocumentStatesAsync(solution, cancellationToken).ConfigureAwait(false);
+
+                if (underlyingGeneratedDocumentStates.Contains(_replacedGeneratedDocumentState.Id))
+                {
+                    // The generated file still exists in the underlying compilation, but the contents may not match the open file if the open file
+                    // is stale. Replace the syntax tree so we have a tree that matches the text.
+                    return underlyingGeneratedDocumentStates.SetState(_replacedGeneratedDocumentState.Id, _replacedGeneratedDocumentState);
+                }
+                else
+                {
+                    // The generated output no longer exists in the underlying compilation. This could happen if the user made
+                    // an edit which would cause this file to no longer exist, but they're still operating on an open representation
+                    // of that file. To ensure that this snapshot is still usable, we'll just add this document back in. This is not a
+                    // semantically correct operation, but working on stale snapshots never has that guarantee.
+                    return underlyingGeneratedDocumentStates.AddRange(ImmutableArray.Create(_replacedGeneratedDocumentState));
+                }
```

Comment:
this pair of options seems to repeat a lot.  makign me think tehre is something weird with our abstractions.  i would have liked it if there was one place where we did the "replace if still exists, else add if not there", and then everything above that transparently picks that up and behaves  normally.,